azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -40,81 +41,243 @@
 
-    def fetch_full_goal_state(self, wire_client):
-        try:
-            logger.info('Fetching goal state [incarnation {0}]', self.incarnation)
+        wire_client.update_host_plugin(container_id, role_config_name)
+
+        return xml_text, xml_doc
+
+    def _fetch_vm_settings(self, force_update=False):
+        """
+        Issues an HTTP request (HostGAPlugin) for the vm settings and returns the response as an ExtensionsGoalStateFromVmSettings.
+        """
+        vm_settings, vm_settings_updated = (None, False)
 
-            xml_text = wire_client.fetch_config(self._hosting_env_uri, wire_client.get_header())
-            self.hosting_env = HostingEnv(xml_text)
+        if conf.get_enable_fast_track():
+            try:
+                timestamp = datetime.datetime.now().isoformat()
 
-            xml_text = wire_client.fetch_config(self._shared_conf_uri, wire_client.get_header())
-            self.shared_conf = SharedConfig(xml_text)
+                vm_settings, vm_settings_updated = self._wire_client.get_host_plugin().fetch_vm_settings(force_update=force_update)
 
-            if self._certs_uri is not None:
-                xml_text = wire_client.fetch_config(self._certs_uri, wire_client.get_header_for_cert())
-                self.certs = Certificates(xml_text)
+            except VmSettingsNotSupported:
+                pass
+            except ResourceGoneError:
+                # retry after refreshing the HostGAPlugin
+                GoalState.update_host_plugin_headers(self._wire_client)
+                vm_settings, vm_settings_updated = self._wire_client.get_host_plugin().fetch_vm_settings(force_update=force_update)
 
-            if self._remote_access_uri is not None:
-                xml_text = wire_client.fetch_config(self._remote_access_uri, wire_client.get_header_for_cert())
-                self.remote_access = RemoteAccess(xml_text)
+            if vm_settings_updated:
+                # The vmSettings are updated independently of the WireServer goal state and they are saved to a separate directory
+                history = GoalStateHistory(timestamp, vm_settings.etag)
+                history.save_vm_settings(vm_settings.get_redacted_text())
 
-            if self._extensions_config_uri is None:
-                self.extensions_config = ExtensionsGoalStateFactory.create_empty()
+        return vm_settings
+
+    def _fetch_extended_goal_state(self, xml_text, xml_doc, vm_settings):
+        """
+        Issues HTTP requests (WireServer) for each of the URIs in the goal state (ExtensionsConfig, Certificate, Remote Access users, etc)
+        and populates the corresponding properties. If the give 'vm_settings' are not None they are used for the extensions goal state,
+        otherwise extensionsConfig is used instead.
+        """
+        try:
+            logger.info('Fetching goal state [incarnation {0}]', self._incarnation)
+
+            self._history.save_goal_state(xml_text)
```

Comment:
now we save the goal state to the history folder as we fetch it, instead of saving to /var/lib/waagent and then flushing it to the history folder when the incarnation changes

Reply:
sample history directory:

```
Length Name
------ ----
       2022-02-03T00_43_06.228674_1    <=== WS goal state
       2022-02-03T00_43_07.544229     <=== status
       2022-02-03T00_44_13.783542_1310808759485798297   <=== HostGAPlugin goal state
       2022-02-03T00_44_13.783997_2
       2022-02-03T00_44_18.097053
       2022-02-03T00_44_48.209362_5216182003277442798
       2022-02-03T00_44_48.209790_3
       2022-02-03T00_44_50.330195
       2022-02-03T00_45_20.442170_4495420063731861706
       2022-02-03T00_45_20.442600_4
       2022-02-03T00_45_24.596102
       2022-02-03T00_45_48.671177_14329981429087073333
       2022-02-03T00_45_48.671655_5
       2022-02-03T00_45_52.837152
       2022-02-03T00_46_22.944478_9343934122256085228
       2022-02-03T00_46_22.944892_6
       2022-02-03T00_46_27.116754
  2993 2022-02-03T00_43_04.659319_1.zip
   785 2022-02-03T00_43_04.724605_12843543198617625598.zip
  2993 2022-02-03T00_43_06.228674_1.zip
   785 2022-02-03T00_43_06.235001_12843543198617625598.zip

```